discovery/hetzner/hetzner.go

```diff
@@ -0,0 +1,173 @@
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/hetznercloud/hcloud-go/hcloud"
+	config_util "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/version"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const (
+	hcloudLabelID          = model.MetaLabelPrefix + "hcloud_id"
+	hcloudLabelName        = model.MetaLabelPrefix + "hcloud_name"
+	hcloudLabelImage       = model.MetaLabelPrefix + "hcloud_image"
+	hcloudLabelPublicIPv4  = model.MetaLabelPrefix + "hcloud_ipv4"
+	hcloudLabelPrivateIPv4 = model.MetaLabelPrefix + "hcloud_private_ipv4"
+	hcloudLabelNetwork     = model.MetaLabelPrefix + "hcloud_network"
+	hcloudLabelLocation    = model.MetaLabelPrefix + "hcloud_location"
+	hcloudLabelDatacenter  = model.MetaLabelPrefix + "hcloud_datacenter"
+	hcloudLabelCPUCores    = model.MetaLabelPrefix + "hcloud_cpu_cores"
+	hcloudLabelCPUType     = model.MetaLabelPrefix + "hcloud_cpu_type"
+	hcloudLabelMemory      = model.MetaLabelPrefix + "hcloud_memory_size"
+	hcloudLabelDisk        = model.MetaLabelPrefix + "hcloud_disk_size"
+	hcloudLabelType        = model.MetaLabelPrefix + "hcloud_server_type"
+)
+
+// DefaultSDConfig is the default Hetzner SD configuration.
+var DefaultSDConfig = SDConfig{
+	Port:            80,
+	RefreshInterval: model.Duration(60 * time.Second),
+	HcloudSDConfig: HcloudSDConfig{
+		Network:  "",
+		Endpoint: hcloud.Endpoint,
+	},
+}
+
+// SDConfig is the configuration for Hetzner based service discovery.
+type SDConfig struct {
+	HTTPClientConfig config_util.HTTPClientConfig `yaml:",inline"`
+
+	RefreshInterval model.Duration `yaml:"refresh_interval"`
+	Port            int            `yaml:"port"`
+	HcloudSDConfig  HcloudSDConfig `yaml:"hcloud"`
+}
+
+// SDConfig is the configuration for Hetzner based service discovery.
+type HcloudSDConfig struct {
+	Network  string `yaml:"network"`
+	Endpoint string `yaml:"endpoint"`
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Discovery periodically performs Hetzner Cloud requests. It implements
+// the Discoverer interface.
+type Discovery struct {
+	*refresh.Discovery
+	client  *hcloud.Client
+	port    int
+	network *hcloud.Network
+}
+
+// NewDiscovery returns a new Discovery which periodically refreshes its targets.
+func NewDiscovery(conf *SDConfig, logger log.Logger) (*Discovery, error) {
+	d := &Discovery{
+		port: conf.Port,
+	}
+
+	rt, err := config_util.NewRoundTripperFromConfig(conf.HTTPClientConfig, "hetzner_sd", false)
+	if err != nil {
+		return nil, err
+	}
+	opts := []hcloud.ClientOption{
+		hcloud.WithToken(string(conf.HTTPClientConfig.BearerToken)),
```

Comment:
Leave this up to the round tripper - HTTPClientConfig is a black box to you, and this is also incorrect for the file.